File: instance_pre-overhead/generate_plot.py
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# Cargar CSV
df = pd.read_csv("result.csv")

# Valores únicos
functions = sorted(df["Functions"].unique())
threads = sorted(df["Threads"].unique())

# Posiciones para las barras
x = np.arange(len(functions))
width = 0.35

# Crear figura
fig, ax = plt.subplots(figsize=(7, 5))

# Colores y hatch para distinguir threads
colors = ["purple", "orange"]
hatches = ["//", ""]

for i, th in enumerate(threads):
    subset = df[df["Threads"] == th].sort_values("Functions")
    ax.bar(
        x + i*width - width/2,
        subset["Time(µs)"],
        width,
        yerr=subset["StdDev"],
        capsize=4,
        label=f"{th} Thread" if th == 1 else f"{th} Threads",
        color=colors[i],
        hatch=hatches[i],
        edgecolor="black"
    )

# Etiquetas y formato
ax.set_xlabel("# of imports")
ax.set_ylabel("Instantiation time (µs)")

# ticks de X en log base 10 (1, 10, 100, 1000...)
ax.set_xticks(x)
ax.set_xticklabels([f"$10^{int(np.log10(f))}$" for f in functions])

ax.set_ylim(0, max(df["Time(µs)"] + df["StdDev"]) * 1.2)

ax.legend()
ax.grid(axis="y", linestyle="--", alpha=0.6)

plt.tight_layout()
plt.savefig("results.png", dpi=300)
plt.show()

File: instance_pre-overhead/generate_plot.m
% === generate_plot.m ===
% Read the CSV file
data = readtable('result.csv');

% Ensure valid column names
data.Properties.VariableNames = matlab.lang.makeValidName(data.Properties.VariableNames);

% Extract unique function counts and thread counts
functions = unique(data.Functions);
threads = unique(data.Threads);

% Build a time matrix
time_matrix = zeros(length(functions), length(threads));
timeVar = data.Properties.VariableNames{contains(data.Properties.VariableNames, 'Time')};

for i = 1:length(functions)
    for j = 1:length(threads)
        time_matrix(i, j) = data.(timeVar)(data.Functions == functions(i) & data.Threads == threads(j));
    end
end

% Create the output folder if it doesn't exist
if ~exist('plots', 'dir')
    mkdir('plots');
end

% Create the figure
figure;
b = bar(log10(functions), time_matrix, 'grouped');
hold on;

% Colors and borders
b(1).FaceColor = [0.35 0.1 0.6]; % purple
b(2).FaceColor = [1 0.75 0];     % yellow
b(1).EdgeColor = 'k';
b(2).EdgeColor = 'k';
b(1).LineStyle = '--';
b(2).LineStyle = '--';

% Axis labels and ticks
xticks(log10(functions));
xticklabels({'10^{0}','10^{1}','10^{2}','10^{3}'});
xlabel('# of imports');
ylabel('Instantiation time (\mus)');
ylim([0 300]);

% Legend
legend({'1 Thread', '16 Threads'}, 'Location', 'northwest');

% Style
set(gca, 'FontSize', 10);
box on;
grid on;
hold off;

% Save plots
saveas(gcf, fullfile('plots', 'instantiation_time.fig'));
saveas(gcf, fullfile('plots', 'instantiation_time.png'));
